---
title: "Covariate analysis of CMC and HBCC matched RNASeq data reprocessed with STAR-feature counts workflow with Ensembl v75 (DLPFC with CQN normlaisation)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./CMC_HBCC_DLPFC_STAR_ENSEMBL_v75_CQN_ModSel.Rmd",
                                 parentId = "syn7501880",
                                 entityName = 'CMC and HBCC - DLPFC - STAR - FEATURE COUNTS - ENSEMBL v75 - CQN')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(ggpubr)

library(reshape2)
library(limma)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)

library(synapseClient)
library(knitr)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')

synapseLogin()

library(doParallel)
library(foreach)

cl = makeCluster(detectCores()-2)
registerDoParallel(cl)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = "syn7501880";
activityName = 'Covariate adjustments';
activityDescription = 'Covariate analysis of STAR aligned feature counts with Ensembl v75 with CQN normalisation (CMC and HBCC DLPFC)';

thisFileName <- 'CMC_HBCC_DLPFC_STAR_ENSEMBL_v75_CQN_ModSel.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Brain_Reg_Net", ref="branch", refName='CMC')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/Rmd/',thisFileName))
```
### Data download
Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download reprocessed counts (CMC and HBCC)
COUNT_CMC_ID = 'syn9874379'
ALL_USED_IDs = COUNT_CMC_ID
COUNT_CMC = downloadFile(COUNT_CMC_ID) 

COUNT_HBCC_ID = 'syn10263483';
ALL_USED_IDs[length(ALL_USED_IDs)+1] = COUNT_HBCC_ID
COUNT_HBCC = downloadFile(COUNT_HBCC_ID)

# Join HBCC and CMC together
COUNT = join_all(list(COUNT_CMC, COUNT_HBCC), type = 'full') %>%
  as.data.frame()

rownames(COUNT) = COUNT$ensembl_gene_id
COUNT$ensembl_gene_id = NULL

# Get covariates (CMC and HBCC)
COVARIATES_CMC_ID = 'syn9874376'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = COVARIATES_CMC_ID
COVARIATES_CMC = downloadFile(COVARIATES_CMC_ID) %>%
  dplyr::mutate(SampleID = colnames(COUNT_CMC)[-(1)],
                LibraryBatch = paste0('CMC',LibraryBatch),
                RibozeroBatch = paste0('CMC',RibozeroBatch),
                FlowcellBatch = paste0('CMC',FlowcellBatch),
                Dx.Tissue = paste(Dx, Tissue, sep=  '_')) %>%
  dplyr::select(-TotalReads, -PercentAligned, -EV.1, -EV.2, -EV.3, -EV.4, -EV.5, -RIN2, -Dx.Tissue.Gender,
                -Dx, -Tissue)

COVARIATES_HBCC_ID = 'syn10263480'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = COVARIATES_HBCC_ID
COVARIATES_HBCC = downloadFile(COVARIATES_HBCC_ID) %>%
  dplyr::mutate(LibraryBatch = paste0('HBCC',LibraryBatch),
                RibozeroBatch = paste0('HBCC',RibozeroBatch),
                FlowcellBatch = paste0('HBCC',FlowcellBatch),
                Institution = 'HBCC',
                Individual_ID = seq(1,length(Institution),1),
                IntronicRate = IntronicRate * 100,
                IntergenicRate = IntergenicRate * 100,
                rRNARate = rRNARate * 100) %>%
  dplyr::rename(MappedReads = Mapped) %>%
  dplyr::select(-TotalReads, -X5prime_Norm, -Mean_Per_Base_Cov., -Mean_CV, -Fragment_Length_StdDev,
                -Mapped_Pairs, -End_1_Mismatch_Rate, -End_2_Mismatch_Rate, -ExonicRate,
                -PercentMapped, -RIN2, -Dx.Tissue.Gender)

COVARIATES = rbindlist(list(CMC = COVARIATES_CMC, 
                            HBCC = COVARIATES_HBCC), 
                       use.names = T, fill = T, idcol = 'Study')
```

### Data preprocessing
Filter DLPFC SCZ and Control samples alone
```{r preprocess.data, fig.height=10, fig.width=15}
# Filter all ACC samples and Other samples
COVARIATES = COVARIATES %>%
  dplyr::filter(Dx.Tissue %in% c("SCZ_DLPFC", "Control_DLPFC"))

COVARIATES %>%
  dplyr::group_by(Dx.Tissue, Study) %>%
  dplyr::summarise(count = n()) %>%
  tidyr::spread(Study, count) %>% 
  kable()

my.theme <- theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1),
                  plot.title=element_text(hjust=0.5))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = RIN, color = Study)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age of Death
p[[2]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = Age_of_Death, color = Study)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeOfDeath') + my.theme

# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = PMI, color = Study)) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI (in hours)') + my.theme

# Intronic Rate
p[[4]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntronicRate, color = Study)) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Intronic Rate') + my.theme

# IntergenicRate
p[[5]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntergenicRate, color = Study)) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Intergenic Rate') + my.theme

# Transcripts Detected
p[[6]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = TranscriptsDetected, color = Study)) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Transcripts Detected') + my.theme

# Mapped Reads
p[[7]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = MappedReads, color = Study)) + geom_boxplot()
p[[7]] = p[[7]] + ggtitle('Mapped Reads') + my.theme

# rRNARate
p[[8]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = rRNARate, color = Study)) + geom_boxplot()
p[[8]] = p[[8]] + ggtitle('rRNARate') + my.theme

p1 = ggpubr::ggarrange(plotlist = p, ncol = 4, nrow = 2) 
p1
```

Age match CMC and HBCC (per Dx.Tissue category)
```{r preprocess.data1}
# Filter all ACC samples and Other samples
max.age = COVARIATES %>%
  dplyr::group_by(Dx.Tissue, Study) %>%
  dplyr::filter(Study == 'CMC') %>%
  dplyr::summarise(mn = quantile(Age_of_Death, 0.5))

COVARIATES = COVARIATES %>%
  dplyr::filter((Dx.Tissue == max.age$Dx.Tissue[1] & Age_of_Death <= max.age$mn[1]) |
                  (Dx.Tissue == max.age$Dx.Tissue[2] & Age_of_Death <= max.age$mn[2]))

COVARIATES %>%
  dplyr::group_by(Dx.Tissue, Study) %>%
  dplyr::summarise(count = n()) %>%
  tidyr::spread(Study, count) %>% 
  kable()
```

PMI match CMC and HBCC (per Dx.Tissue category)
```{r preprocess.data2}
max.pmi = COVARIATES %>%
  dplyr::group_by(Dx.Tissue) %>%
  dplyr::filter(Study == 'HBCC') %>%
  dplyr::summarise(mn = quantile(PMI, 0.5))

COVARIATES = COVARIATES %>%
  dplyr::filter((Dx.Tissue == max.age$Dx.Tissue[1] & PMI <= max.pmi$mn[1]) |
                  (Dx.Tissue == max.age$Dx.Tissue[2] & PMI <= max.pmi$mn[2]))

COVARIATES %>%
  dplyr::group_by(Dx.Tissue, Study) %>%
  dplyr::summarise(count = n()) %>%
  tidyr::spread(Study, count) %>% 
  kable()
```

```{r gene.param}
## Get GC content from biomart
backgroundGenes = data.frame(ensembl_gene_id = rownames(COUNT))

# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "dec2016.archive.ensembl.org", dataset = "hsapiens_gene_ensembl")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "percentage_gc_content", "gene_biotype"),
                       filters = "ensembl_gene_id", values = backgroundGenes$ensembl_gene_id,
                       mart = mart)

GENE.GC.CONT = Ensemble2HGNC %>%
  dplyr::select(ensembl_gene_id, percentage_gc_content) %>%
  unique
rownames(GENE.GC.CONT) = GENE.GC.CONT$ensembl_gene_id

# Get gene lengths from synapse
GENE.LEN = downloadFile('syn8414424') %>%
  dplyr::select(Geneid, Length) %>%
  dplyr::filter(Geneid %in% rownames(COUNT)) %>%
  unique() %>%
  dplyr::rename(ensembl_gene_id = Geneid)
rownames(GENE.LEN) = GENE.LEN$ensembl_gene_id
```

### Covariates clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c('Study', 'Individual_ID', 'Institution', 'Gender', 'LibraryBatch', 'Dx.Tissue', 'RibozeroBatch',
                      'FlowcellBatch')
ContCovariates <- c("Age_of_Death", "PMI", "RIN", "MappedReads", "IntragenicRate", "IntronicRate", "IntergenicRate",
                    "GenesDetected", "TranscriptsDetected", "ExpProfEfficiency", "rRNARate")

# Find inter relation between factor covariates
rownames(COVARIATES) = COVARIATES$SampleID
COVARIATES = COVARIATES[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES[,FactorCovariates] <- data.frame(lapply(COVARIATES[,FactorCovariates],function(x){
  x <- sapply(x,function(y){str_replace_all(as.character(y),'[^[:alnum:]]','_')})}))

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], function(x){
  x = as.numeric(as.character(gsub('[\\,\\%]','',x)))
})

# Add in RIN^2 values
COVARIATES$RIN2 = COVARIATES$RIN^2
ContCovariates = c(ContCovariates, 'RIN2')
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, fig.width=10, fig.height=8}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
tmp = COVARIATES.CORRELATION$ESTIMATE
tmp[COVARIATES.CORRELATION$PVAL > 0.05] = 0
h = Heatmap(tmp, col = colorRamp2(c(-1,0,1), c('blue','white','red')), name = 'AssocEstimate')
draw(h, heatmap_legend_side = 'left')
```

### Explore metatdata
```{r data.explore, fig.width = 15, fig.height = 10}
my.theme <- theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1), plot.title=element_text(hjust=0.5))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = RIN, color = Study)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age of Death
p[[2]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = Age_of_Death, color = Study)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeOfDeath') + my.theme

# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = PMI, color = Study)) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI (in hours)') + my.theme

# Intronic Rate
p[[4]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntronicRate, color = Study)) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Intronic Rate') + my.theme

# IntergenicRate
p[[5]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntergenicRate, color = Study)) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Intergenic Rate') + my.theme

# Transcripts Detected
p[[6]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = TranscriptsDetected, color = Study)) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Transcripts Detected') + my.theme

# Mapped Reads
p[[7]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = MappedReads, color = Study)) + geom_boxplot()
p[[7]] = p[[7]] + ggtitle('Mapped Reads') + my.theme

# rRNARate
p[[8]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = rRNARate, color = Study)) + geom_boxplot()
p[[8]] = p[[8]] + ggtitle('rRNARate') + my.theme

p1 = ggpubr::ggarrange(plotlist = p, ncol = 4, nrow = 2)
p1
```
### Filter genes
Remove genes that have less than 1 cpm counts in at least 50% of samples per Dx.Tissue. Also remove genes with missing gene length and percentage GC content
```{r cpmnormalisation}
COUNT[is.na(COUNT)] = 0
genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('SampleID') %>%
  dlply(.(Dx.Tissue), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    processed.counts$filteredExprMatrix$genes
  }, COUNT)

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC.CONT$ensembl_gene_id[!is.na(GENE.GC.CONT$percentage_gc_content)]) %>%
  intersect(GENE.LEN$ensembl_gene_id[!is.na(GENE.LEN$Length)])

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, intersect(colnames(COUNT), rownames(COVARIATES))],
                                                 MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
pct.pc = PROCESSED_COUNTS$filteredExprMatrix$genes %>%
  dplyr::rename(ensembl_gene_id = genes) %>%
  left_join(Ensemble2HGNC) %>%
  group_by(gene_biotype) %>%
  summarise(fraction  = n()) %>%
  filter(fraction > 100) %>%
  dplyr::mutate(fraction = fraction/length(PROCESSED_COUNTS$filteredExprMatrix$genes[,1]))
```
`r dim(PROCESSED_COUNTS$filteredExprMatrix$genes)[1]` genes are used for the analysis. 

Following is the distribution of biotypes
`r kable(pct.pc)`

### Outlier analysis
Detect outlier samples based on their expression (logCPM) pattern
```{r detect.outliers, fig.height=8, fig.width=8, results='asis', cache = FALSE}
# Find principal components of expression to plot
PC <- prcomp(voom(PROCESSED_COUNTS$filteredExprMatrix$counts)$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_') %>%
  dplyr::mutate(label = SampleID)
plotdata$label[!(plotdata$SampleID %in% indToRemove)] = ''

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=Age_of_Death))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~.)
# p <- p + geom_text(aes(label= label), size=4, hjust=0)
p

tmp = tidyr::separate(COVARIATES, Dx.Tissue, c('Dx', 'Tissue'), sep = '_') %>%
  group_by(Study, Dx, Tissue) %>%
  summarise(count = n()) %>%
  spread(Dx, count)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples from `r length(unique(COVARIATES$Individual_ID))` unique individuals

Distribution of samples are:
  `r kable(tmp)`

### Winsorise counts (Gene outliers in samples)
```{r winsorise.data}
LOG.CPM = cpm(PROCESSED_COUNTS$filteredExprMatrix$counts, log = T)

# Set gene counts in specific samples that are deviating 3 sd from other samples to 3SD limit
LOG.CPM = apply(LOG.CPM, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  
  ind.low = which(x == max(x[which(x <= mn-3*std.dev)], na.rm = T))
  ind.high = which(x == min(x[which(x >= mn+3*std.dev)], na.rm = T))
  
  x[x < (mn-3*std.dev)] = x[ind.low]
  x[x > (mn+3*std.dev)] = x[ind.high]
  return(x)
}) %>% t

# Back calculate counts for further analysis
LIB.SIZE = colSums(PROCESSED_COUNTS$filteredExprMatrix$counts)
NEW.COUNTS = (2^LOG.CPM) * t(replicate(dim(LOG.CPM)[1], LIB.SIZE))/1e6
```

### Library Normalisation
Initial library normalisation usign cqn
```{r cqn}
# Compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(NEW.COUNTS, 
                          x = GENE.GC.CONT[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'percentage_gc_content'],
                          lengths = GENE.LEN[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'Length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset

# Set gene counts in specific samples that are deviating 3 sd from other samples to 3SD limit
CQN.GENE_EXPRESSION$E = apply(CQN.GENE_EXPRESSION$E, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  
  ind.low = which(x == max(x[which(x <= mn-3*std.dev)], na.rm = T))
  ind.high = which(x == min(x[which(x >= mn+3*std.dev)], na.rm = T))
  
  x[x < (mn-3*std.dev)] = x[ind.low]
  x[x > (mn+3*std.dev)] = x[ind.high]
  return(x)
}) %>% t

# Back calculate counts for further analysis
LIB.SIZE = colSums(PROCESSED_COUNTS$filteredExprMatrix$counts)
NEW.COUNTS = (2^CQN.GENE_EXPRESSION$E) * t(replicate(dim(CQN.GENE_EXPRESSION$E)[1], LIB.SIZE))/1e6
```

### Co-expression distribution
Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$E))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Sample clustering
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis', cache=FALSE}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_')

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=Age_of_Death))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~.)
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p
```
Tree based clustering of samples
```{r decompse.normalise.data.1, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c("Institution", "Gender", "LibraryBatch", "Dx.Tissue")])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```
```{r temp, include = F}
dev.off()
gc()
```

### Distribution of samples (log cpm)
```{r lcpm.dist, cache=FALSE}
# Plot abberent distribution of logcpm counts
tmp1 = (CQN.GENE_EXPRESSION$E) %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  tidyr::gather(SampleID, logCPM, -ensembl_gene_id) %>%
  left_join(COVARIATES %>%
              rownameToFirstColumn('SampleID') %>%
              tidyr::separate(Dx.Tissue, c('Dx', 'Tissue'), sep = '_'))

p = ggplot(tmp1, aes(x = logCPM, color = SampleID)) + geom_density() 
p = p + theme(legend.position = 'NONE') + facet_grid(Dx+.~Tissue, scale = 'free')
p
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$E, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Iterative Normalisation
Dx.Tissue.Study is chosen as the primary variable of interest (i.e., covariate selection is conditioned on diagnosis and tissue)
```{r iterative.normalisation}
# Primary variable of interest
COVARIATES$Dx.Tissue.Study = paste(COVARIATES$Dx.Tissue, COVARIATES$Study, sep = '_') %>%
  factor()
postAdjustCovars = c('Dx.Tissue.Study');
avoid.covariates = c('Dx.Tissue', 'Study', 'Individual_ID')

# Assign residual covariates
residualCovars = setdiff(preAdjustedSigCovars$significantCovars, 
                         c(postAdjustCovars, avoid.covariates))
residualSigCovars = preAdjustedSigCovars
covariatesEffects = preAdjustedSigCovars$Effects.significantCovars[residualCovars]
postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects))) %>% unique()

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 20){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', ')))

  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]

  # Estimate voom weights for dispersion control
  VOOM.GENE_EXPRESSION = voom(NEW.COUNTS, design=DM1$design, plot=F)
  VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E
  
  # Fit linear model using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(CQN.GENE_EXPRESSION$E,
                            design = DM1$design,
                            weights = VOOM.GENE_EXPRESSION$weights)

  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                                CQN.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates),
                         c(postAdjustCovars, avoid.covariates))

  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)

  # Add postadjusted covariates (if any)
  residCovars = setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, avoid.covariates))
  covariatesEffects = residualSigCovars$Effects.significantCovars[residCovars]

  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}

modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```

### Normalisation
```{r custom.normalisation, results='asis'}
writeLines(paste('Using following covariates in the model:',
                 paste(postAdjustCovars, collapse=', ')))

# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)$design
DESIGN = DESIGN[,linColumnFinder(DESIGN)$indepCols]

# Estimate voom weights
VOOM.GENE_EXPRESSION = voom(NEW.COUNTS, design = DESIGN)
VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E

# Fit linear model using new weights and new design
ADJUSTED.FIT = lmFit(VOOM.GENE_EXPRESSION)

# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(ADJUSTED.FIT, CQN.GENE_EXPRESSION$E)
```

### Residual calculation
Calculate weighted residuals and add back "Dx.Tissue.Study" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = grep("Dx.Tissue.Study", colnames(DESIGN), value = T)
RESIDUAL.GENE_EXPRESSION = RESIDUAL.GENE_EXPRESSION +
  ADJUSTED.FIT$coefficients[,varsToAddIn] %*% t(DESIGN[,varsToAddIn])
```
Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 4 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_') %>%
  dplyr::mutate(label = SampleID)
# plotdata$label[!(plotdata$SampleID %in% top.nsamples)] = ''

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=Age_of_Death))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~.)
# p <- p + geom_text(aes(label= label), size=4, hjust=0)
p
```

```{r decompse.normalise.data2.1, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Dx.Tissue','Institution')])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(RESIDUAL.GENE_EXPRESSION)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Dx.Tissue','Institution')]))
```

```{r temp1, include=F}
dev.off()
gc()
```

### Differential expression analysis (SCZ - Control)
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp, fig.height=15, fig.width=20}
# Fit contrast
contrast = makeContrasts(contrasts=c("Dx.Tissue.StudySCZ_DLPFC_CMC-Dx.Tissue.StudyControl_DLPFC_CMC",
                                     "Dx.Tissue.StudySCZ_DLPFC_HBCC-Dx.Tissue.StudyControl_DLPFC_HBCC"),
                         levels = colnames(ADJUSTED.FIT$coefficients))
FIT.CONTR = contrasts.fit(ADJUSTED.FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:dim(contrast)[2], function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = TRUE) %>%
    rownameToFirstColumn('ensembl_gene_id') 
}, FIT.CONTR) 
names(DE) = gsub('Dx.Tissue.Study', '', colnames(contrast))

DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Dx.Tissue','',Comparison),
                Direction = sign(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  left_join(Ensemble2HGNC) %>%
  left_join(GENE.LEN)
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison, logFC) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_DOWN = length(unique(ensembl_gene_id[logFC < 0])),
                   FDR_0_05_UP = length(unique(ensembl_gene_id[logFC > 0])))

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_FC_1.2_DOWN = length(unique(ensembl_gene_id[logFC < 0])),
                   FDR_0_05_FC_1.2_UP = length(unique(ensembl_gene_id[logFC > 0])))

kable(full_join(tmp,tmp1))

p = DE %>%
  ggplot(aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position = 'top')
p = p + facet_grid(Comparison~., scales = 'free_y')
p = p + ggtitle('Differential Expression between Schizophrenia and Control')
p

all.diff.exp = list(Diagnosis.Study = DE)
all.fit = list(Diagnosis.Study = ADJUSTED.FIT)
```

### Associate differential expression results with gc content, gene length and average expression
```{r associate.de, fig.height=10, fig.width=15}
pl = list()
pl[[1]] = ggplot(DE, aes(x = log10(Length), y = logFC, color = Direction)) + geom_point() 
pl[[1]] = pl[[1]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = log10(Length), y = logFC))
pl[[1]] = pl[[1]] + facet_grid(Comparison~.) + scale_color_manual(values = c('green','grey','red'))
pl[[1]] = pl[[1]] + theme(legend.position = 'top')

pl[[2]] = ggplot(DE, aes(x = percentage_gc_content, y = logFC, color = Direction)) + geom_point()
pl[[2]] = pl[[2]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = percentage_gc_content, y = logFC))
pl[[2]] = pl[[2]] + facet_grid(Comparison~.) + scale_color_manual(values = c('green','grey','red'))
pl[[2]] = pl[[2]] + theme(legend.position = 'top')

pl[[3]] = ggplot(DE, aes(x = AveExpr, y = logFC, color = Direction)) + geom_point()
pl[[3]] = pl[[3]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = AveExpr, y = logFC))
pl[[3]] = pl[[3]] + facet_grid(Comparison~.) + scale_color_manual(values = c('green','grey','red'))
pl[[3]] = pl[[3]] + theme(legend.position = 'top')

multiplot(plotlist = pl, cols = 3)
```

### Differential expression analysis (SCZ - Control gender specific)
Genes that are differentially expressed at an FDR <= 0.05 are
```{r diffExp1, fig.height=15, fig.width=20}
COVARIATES$Dx.Tissue.Study.Gender = paste(COVARIATES$Dx.Tissue.Study,
                                          COVARIATES$Gender,
                                          sep = '_') %>%
  factor()

postAdjustCovars = c('Dx.Tissue.Study.Gender', setdiff(postAdjustCovars, c('Dx.Tissue.Study', 'Gender')))
writeLines(paste('Using following covariates in the model:', paste(postAdjustCovars, collapse=', ')))

# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)$design
DESIGN = DESIGN[,linColumnFinder(DESIGN)$indepCols]

# Estimate voom weights
VOOM.GENE_EXPRESSION = voom(NEW.COUNTS, design = DESIGN)
VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E

# Fit linear model using new weights and new design
ADJUSTED.FIT = lmFit(VOOM.GENE_EXPRESSION)

# Fit contrast
contrast = makeContrasts(contrasts=
                           c("Dx.Tissue.Study.GenderSCZ_DLPFC_CMC_Female-Dx.Tissue.Study.GenderControl_DLPFC_CMC_Female",
                             "Dx.Tissue.Study.GenderSCZ_DLPFC_CMC_Male-Dx.Tissue.Study.GenderControl_DLPFC_CMC_Male",
                             "Dx.Tissue.Study.GenderSCZ_DLPFC_HBCC_Female-Dx.Tissue.Study.GenderControl_DLPFC_HBCC_Female",
                             "Dx.Tissue.Study.GenderSCZ_DLPFC_HBCC_Male-Dx.Tissue.Study.GenderControl_DLPFC_HBCC_Male"),
                         levels = colnames(ADJUSTED.FIT$coefficients))
FIT.CONTR = contrasts.fit(ADJUSTED.FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:dim(contrast)[2], function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = TRUE) %>%
    rownameToFirstColumn('ensembl_gene_id') 
}, FIT.CONTR) 
names(DE) = gsub('Dx.Tissue.Study.Gender', '', colnames(contrast))

DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Direction = sign(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  left_join(Ensemble2HGNC) %>%
  left_join(GENE.LEN)
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison, logFC) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_DOWN = length(unique(ensembl_gene_id[logFC < 0])),
                   FDR_0_05_UP = length(unique(ensembl_gene_id[logFC > 0])))

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_FC_1.2_DOWN = length(unique(ensembl_gene_id[logFC < 0])),
                   FDR_0_05_FC_1.2_UP = length(unique(ensembl_gene_id[logFC > 0])))

kable(full_join(tmp,tmp1))

p = DE %>%
  ggplot(aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position = 'top')
p = p + facet_grid(Comparison~., scales = 'free_y')
p = p + ggtitle('Differential Expression between Schizophrenia and Control')
p

all.diff.exp = c(all.diff.exp, list(Diagnosis.Study.Gender = DE))
all.fit = c(all.fit, list(Diagnosis.Study.Gender = ADJUSTED.FIT))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Set annotations
all.annotations = list(
dataType = 'mRNA',
dataSubType = 'geneExp',
summaryLevel = 'gene',
assay	 = 'RNAseq',

tissueTypeAbrv	= 'DLPFC', 
study = 'CMC and HBCC', 

organism = 'HomoSapiens',
consortium	= 'CMC',

normalizationStatus	= TRUE,

normalizationType	= 'CQN'
)

# Code
CODE <- Folder(name = "CMC and HBCC - DLPFC - STAR - FEATURE COUNTS - ENSEMBL v75 - CQN", parentId = parentId)
annotations(CODE) = all.annotations
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'CMC_HBCC_DLPFC_Covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('CMC_HBCC_DLPFC_Covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
annotations(COV_OBJ) = annotations(CODE)
annotations(COV_OBJ)$dataSubType = 'covariates'
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
executed = thisFile, activityDescription = activityDescription)

# Store filtered counts
PROCESSED_COUNTS$filteredExprMatrix$counts %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_HBCC_DLPFC_Counts.tsv', sep = '\t', row.names=F, quote=F)
COUNT_OBJ = File('CMC_HBCC_DLPFC_Counts.tsv', name = 'Counts (filtered raw)', parentId = CODE$properties$id)
annotations(COUNT_OBJ) = annotations(CODE)
annotations(COUNT_OBJ)$dataSubType = 'filteredCounts'
COUNT_OBJ = synStore(COUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store logCPM
CQN.GENE_EXPRESSION$y[is.na(CQN.GENE_EXPRESSION$E)] = NA
CQN.GENE_EXPRESSION$y %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_HBCC_DLPFC_logCPM.tsv', sep = '\t', row.names=F, quote=F)
LCOUNT_OBJ = File('CMC_HBCC_DLPFC_logCPM.tsv', name = 'Counts (filtered logCPM)', parentId = CODE$properties$id)
annotations(LCOUNT_OBJ) = annotations(CODE)
annotations(LCOUNT_OBJ)$dataSubType = 'filteredLCPM'
LCOUNT_OBJ = synStore(LCOUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store cqn offsets
CQN.GENE_EXPRESSION$offset[is.na(CQN.GENE_EXPRESSION$E)] = NA
CQN.GENE_EXPRESSION$offset %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_HBCC_DLPFC_offset.tsv', sep = '\t', row.names=F, quote=F)
OFFSET_OBJ = File('CMC_HBCC_DLPFC_offset.tsv', name = 'Gene length and GC content offset', parentId = CODE$properties$id)
annotations(OFFSET_OBJ) = annotations(CODE)
annotations(OFFSET_OBJ)$dataSubType = 'offset'
OFFSET_OBJ = synStore(OFFSET_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store design matrix
DESIGN %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'CMC_HBCC_DLPFC_Design.tsv', sep = '\t', row.names=F, quote=F)
DM_OBJ = File('CMC_HBCC_DLPFC_Design.tsv', name = 'Design Matrix', parentId = CODE$properties$id)
annotations(DM_OBJ) = annotations(CODE)
annotations(DM_OBJ)$dataSubType = 'designMatrix'
DM_OBJ = synStore(DM_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store differential expression results
all.diff.exp %>%
  rbindlist(use.names = T, fill = T) %>%
  write.table(file = 'CMC_HBCC_DLPFC_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CMC_HBCC_DLPFC_DiffExpression.tsv', 
name = 'Differential Expression Results (Dx.Tissue)', 
parentId = CODE$properties$id)
annotations(DEXP_OBJ) = annotations(CODE)
annotations(DEXP_OBJ)$dataSubType = 'diffExp'
DEXP_OBJ = synStore(DEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))
```
|  *Results*                                  |  *SynapseID*                     |
|  -----------------------------------------  |   ---------                      |
|  Covariates                                 |  `r COV_OBJ$properties$id`       |
|  Counts (raw)                               |  `r COUNT_OBJ$properties$id`     |
|  Counts (lcpm)                              |  `r LCOUNT_OBJ$properties$id`    |
|  Offset (for gene length and gc content)    |  `r OFFSET_OBJ$properties$id`    |
|  Design Matrix                              |  `r DM_OBJ$properties$id`        |
|  Residual Expression (for network analysis) |  `r nEXP_OBJ$properties$id`      |
|  Differential Expression                    |  `r DEXP_OBJ$properties$id`      |
|  Enrichment Analysis                        |  `r ENRICH_OBJ$properties$id`    |

### Source Code
[R markdown](`r thisFile`)